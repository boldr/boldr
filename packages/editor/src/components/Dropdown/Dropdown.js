/* eslint-disable react/no-array-index-key, react/no-unused-prop-types, no-inline-comments */
// @flow
import React from 'react';
import type { Node } from 'react';
import cn from 'classnames';
import { stopPropagation } from '../../utils/common';

export type Props = {
  children: Node,
  onChange?: Function,
  className?: string,
  title: string,
  expanded?: boolean,
  doExpand: Function,
  doCollapse: Function,
  onExpandEvent: Function,
  isSkinny: boolean,
  ariaLabel?: string,
  optionWrapperClassName?: string,
};

type State = {
  highlighted: number,
};

export default class Dropdown extends React.Component<Props, State> {
  static defaultProps = {
    isSkinny: false,
  };

  state: State = {
    highlighted: -1,
  };

  componentWillReceiveProps(props: Props) {
    if (this.props.expanded && !props.expanded) {
      this.setState({
        highlighted: -1,
      });
    }
  }

  props: Props;

  onChange: Function = (value: any): void => {
    const { onChange } = this.props;
    if (onChange) {
      onChange(value);
    }
    this.toggleExpansion();
  };

  setHighlighted: Function = (highlighted: number): void => {
    this.setState({
      highlighted,
    });
  };

  toggleExpansion: Function = (): void => {
    const { doExpand, doCollapse, expanded } = this.props;
    if (expanded) {
      doCollapse();
    } else {
      doExpand();
    }
  };

  render() {
    const {
      expanded,
      children,
      className,
      optionWrapperClassName,
      ariaLabel,
      onExpandEvent,
      title,
    } = this.props;
    const { highlighted } = this.state;
    // $FlowIssue
    const options: Array<Node> = children.slice(1, children.length);
    return (
      <div
        className={cn(
          'be-dd__wrap',
          {
            'is-skinny': this.props.isSkinny,
          },
          className,
        )}
        aria-expanded={expanded}
        aria-label={ariaLabel || 'be-dropdown'}>
        <a className={cn('be-dd__selected-txt')} onClick={onExpandEvent} title={title}>
          {/* // $FlowIssue */}
          {children[0]}
          <div
            className={cn({
              'be-caret--closed': expanded,
              'be-caret--open': !expanded,
            })}
          />
        </a>
        {expanded ? (
          <ul className={cn('be-dd__opt-wrap', optionWrapperClassName)} onClick={stopPropagation}>
            {React.Children.map(options, (option, index) => {
              const temp =
                option &&
                // $FlowIssue
                React.cloneElement(option, {
                  onSelect: this.onChange,
                  highlighted: highlighted === index,
                  setHighlighted: this.setHighlighted,
                  index,
                });
              return temp;
            })}
          </ul>
        ) : (
          undefined
        )}
      </div>
    );
  }
}
