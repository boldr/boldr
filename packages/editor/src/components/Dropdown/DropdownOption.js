/* eslint-disable react/no-unused-prop-types */
/* @flow */

import React from 'react';
import type { Node } from 'react';
import cn from 'classnames';

export type Props = {
  children: Node,
  value?: any,
  onClick?: Function,
  onSelect?: Function,
  setHighlighted?: Function,
  index?: number,
  disabled?: boolean,
  active?: boolean,
  highlighted?: boolean,
  className?: string,
  title?: string,
};

export default class DropdownOption extends React.PureComponent<Props, *> {
  props: Props;

  onClick: Function = (event: SyntheticEvent<>): void => {
    const { onSelect, onClick, value, disabled } = this.props;
    if (!disabled) {
      if (onSelect) {
        onSelect(value);
      }
      if (onClick) {
        event.stopPropagation();
        onClick(value);
      }
    }
  };

  setHighlighted: Function = (): void => {
    const { setHighlighted, index } = this.props;
    setHighlighted(index);
  };

  resetHighlighted: Function = (): void => {
    const { setHighlighted } = this.props;
    setHighlighted(-1);
  };

  render(): Node {
    const { children, active, highlighted, title, disabled } = this.props;
    return (
      <li
        className={cn('be-dd__opt--default', {
          'be-dd__opt--active': active,
          'be-dd__opt--highlight': highlighted,
          'be-dd__opt--disabled': disabled,
        })}
        onMouseEnter={this.setHighlighted}
        onMouseLeave={this.resetHighlighted}
        onClick={this.onClick}
        onKeyDown={this.onClick}
        aria-selected={active}
        title={title}>
        {children}
      </li>
    );
  }
}
