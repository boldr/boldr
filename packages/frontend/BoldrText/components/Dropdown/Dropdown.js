/* eslint-disable no-param-reassign, no-unused-expressions, no-return-assign */
// @flow weak
import React, { Component } from 'react';
import uniqueId from 'lodash/uniqueId';
import cN from 'classnames';
import ResponsiveHelper from '../../core/helpers/responsive';

export type Props = {
  caption: string,
  hoverTitle: string,
  disabled: boolean,
  showDropDownArrow: boolean,
  arrowActive: boolean,
  className?: string,
  children: ReactChildren,
  hideOnBlur: boolean,
  componentId: string,
};
type State = {
  active: boolean,
  offset: number,
};

export default class Dropdown extends Component<Props, State> {
  constructor(props: Props) {
    super();
    (this: any).componentId = props.componentId || `BE-DROPDOWN-${uniqueId()}`;
    (this: any).alive = false;
    (this: any).responsiveResolveId = null;
    (this: any).dropDownHandlerElement = null;
    (this: any).dropDownContentElement = null;
    this.state = {
      active: false,
      offset: 0,
    };
  }
  state: State;
  componentDidMount() {
    // $FlowIssue
    this.alive = true;
    this.fixDropDownPosition();
    // $FlowIssue
    document.body.addEventListener('click', event => {
      this.registerClickEvent(event);
    });
    // $FlowIssue
    this.responsiveResolveId = ResponsiveHelper.resolve(() => {
      this.fixDropDownPosition();
    });
  }

  componentWillReceiveProps(next: any) {
    if (!this.props.disabled && next.disabled) {
      this.hide();
    }
  }

  componentWillUnmount() {
    // $FlowIssue
    document.body.removeEventListener('click', (event: MouseEvent) => {
      this.registerClickEvent(event);
    });
    // $FlowIssue
    this.alive = false;
    // $FlowIssue
    ResponsiveHelper.unresolve(this.responsiveResolveId);
  }

  props: Props;

  fixDropDownPosition() {
    let offset = 0;
    // $FlowIssue
    const viewWidth = document.body.getBoundingClientRect().width;
    // $FlowIssue
    const handlerRect = this.dropDownHandlerElement.getBoundingClientRect();
    // $FlowIssue
    const contentRect = this.dropDownContentElement.getBoundingClientRect();
    let right = handlerRect.right - handlerRect.width / 2 + contentRect.width / 2;
    const left = handlerRect.left + handlerRect.width / 2 - contentRect.width / 2;

    right = viewWidth - right;

    if (right < 10) {
      offset = right - 10;
    } else if (left < 10) {
      offset = left * -1 + 10;
    }

    offset !== this.state.offset && this.setState({ offset });
  }

  registerClickEvent(event: MouseEvent) {
    const { hideOnBlur } = this.props;
    let active = false;
    // $FlowIssue
    if (event.target.dataset.beComponentId === this.componentId) {
      active = event.target.dataset.keepActive ? true : !this.state.active;
    } else if (hideOnBlur === false) {
      active = this.state.active;
    }

    this.alive && this.setState({ active });
  }

  show() {
    this.setState({
      active: true,
    });
  }

  hide() {
    this.setState({
      active: false,
    });
  }
  render() {
    let { active, offset } = this.state;
    const {
      caption,
      hoverTitle,
      disabled,
      showDropDownArrow,
      arrowActive,
      className,
      children,
    } = this.props;

    disabled && (active = false);
    const arrowClassName = cN('be-dropdown__arrow', {
      active: arrowActive,
    });
    const ddClassName = cN(
      'be-dropdown',
      {
        // eslint-disable-next-line
        active: active,
        // eslint-disable-next-line
        disabled: disabled,
      },
      className,
    );
    return (
      // $FlowIssue
      <div id={this.componentId} className={ddClassName}>
        <button
          className={cN('be-dropdown__handler')}
          title={hoverTitle}
          data-be-component-id={(this: any).componentId}
          ref={instance => ((this: any).dropDownHandlerElement = instance)}
        >
          <span className="be-dropdown__title">
            {caption}
          </span>
          {showDropDownArrow !== false && <i className={cN('icon-drop-down')} />}
        </button>
        <div
          className={cN('be-dropdown__content')}
          style={{ marginLeft: `${offset}px` }}
          ref={instance => ((this: any).dropDownContentElement = instance)}
        >
          <i style={{ marginLeft: offset * -1 }} className={arrowClassName} />
          {children}
        </div>
      </div>
    );
  }
}
