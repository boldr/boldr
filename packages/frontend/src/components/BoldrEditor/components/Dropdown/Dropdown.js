/* @flow */

import React, { Component } from 'react';
import classNames from 'classnames';

import styled from 'styled-components';

const Holder = styled.div`
  z-index: 190;
  background: #fff;
  border: 1px solid #eaeaea;
`;
export type Props = {
  children: Array<ReactChildren>,
  onChange: ?Function,
  className: ?string,
  title: string,
  onExpandEvent: ?Function,
  optionWrapperClassName: ?string,
};

export default class Dropdown extends Component {
  static defaultProps = {
    title: '',
  };

  state: Object = {
    expanded: false,
    highlighted: -1,
  };

  props: Props;
  handleChange: Function = (event: Event): void => {
    const value: string = event.target.value;
    this.props.onChange(value);
  };

  setHighlighted: Function = (highlighted: number): void => {
    this.setState({
      highlighted,
    });
  };
  expandMenu = () => {
    this.setState({
      expanded: true,
    });
  };
  collapseMenu = () => {
    this.setState({
      expanded: false,
    });
  };
  toggleExpansion: Function = (): void => {
    // const { doExpand, doCollapse, expanded } = this.props;
    if (this.state.expanded) {
      this.collapseMenu();
    } else {
      this.expandMenu();
    }
  };

  render() {
    const { children, className } = this.props;
    const { highlighted } = this.state;
    const { expanded } = this.state;
    const options = children.slice(1, children.length);
    return (
      <div
        className={classNames('boldredit-dropdown__wrapper', className)}
        aria-expanded={expanded}
        aria-label="boldredit-dropdown-control"
      >
        <a
          className="boldredit-dropdown__selectedtext"
          onClick={this.toggleExpansion}
          title={this.props.title}
        >
          {children[0]}
          <div
            className={classNames({
              'boldredit-dropdown__carettoclose': expanded,
              'boldredit-dropdown__carettoopen': !expanded,
            })}
          />
        </a>
        {this.state.expanded
          ? <Holder>
              {React.Children.map(options, (option, index) => {
                const temp =
                  option &&
                  React.cloneElement(option, {
                    onSelect: this.props.onChange,
                    highlighted: highlighted === index,
                    setHighlighted: this.setHighlighted,
                    index,
                  });
                return temp;
              })}
            </Holder>
          : null}
      </div>
    );
  }
}
