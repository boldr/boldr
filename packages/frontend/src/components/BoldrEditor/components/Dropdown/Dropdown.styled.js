import styled from 'styled-components';

export const DropdownWrapper = styled.div`
  height: 30px;
  background: #fff;
  border: 1px solid #e5e5e5;
  border-radius: 2px;
  margin: 0 3px;
  width: ${props => (props.isSkinny ? '60px' : '120px')};
  text-transform: capitalize;
  cursor: pointer;
  a {
    color: rgba(0, 0, 0, 0.87);
  }

  &:focus {
    outline: none;
  }

  &:hover {
    background-color: #fff;
    box-shadow: 1px 1px 0px #bfbdbd;
  }

  &:active {
    box-shadow: 1px 1px 0px #bfbdbd inset;
  }
`;
export const DropdownOptionWrapper = styled.ul`
  position: relative;
  z-index: 100;
  overflow-y: scroll;
  width: 98%;
  max-height: 250px;
  padding: 0;
  background: #fff;
  border: 1px solid #f1f1f1;
  border-radius: 2px;
  margin: 0;
`;
export const DropdownSelectedText = styled.a`
  position: relative;
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0 5px;
`;

export const CaretOpen = styled.div`
  position: absolute;
  top: 35%;
  right: 10%;
  width: 0;
  height: 0;
  border-top: 6px solid black;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
`;
export const CaretClosed = styled.div`
  position: absolute;
  top: 35%;
  right: 10%;
  width: 0;
  height: 0;
  border-right: 5px solid transparent;
  border-bottom: 6px solid black;
  border-left: 5px solid transparent;
`;
