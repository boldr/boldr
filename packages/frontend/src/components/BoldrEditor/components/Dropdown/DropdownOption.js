/* eslint-disable react/no-unused-prop-types */
/* @flow */

import * as React from 'react';
import styled from 'styled-components';

export type Props = {
  children: React.ChildrenArray<React.Node>,
  value?: any,
  onClick?: Function,
  onSelect?: Function,
  setHighlighted?: Function,
  index?: number,
  disabled?: boolean,
  active?: boolean,
  highlighted?: boolean,
  className?: string,
  title?: string,
};

const DropdownOpt = styled.li`
  display: flex;
  align-items: center;
  min-height: 25px;
  padding: 0 5px;
  background: ${props => (props.highlighted ? '#f1f1f1' : '#fff')};
`;

export default class DropdownOption extends React.Component<Props, *> {
  props: Props;

  onClick: Function = (event): void => {
    const { onSelect, onClick, value, disabled } = this.props;
    if (!disabled) {
      if (onSelect) {
        onSelect(value);
      }
      if (onClick) {
        event.stopPropagation();
        onClick(value);
      }
    }
  };

  setHighlighted: Function = (): void => {
    const { setHighlighted, index } = this.props;
    setHighlighted(index);
  };

  resetHighlighted: Function = (): void => {
    const { setHighlighted } = this.props;
    setHighlighted(-1);
  };

  render(): React.Node {
    const { children, active, highlighted, title } = this.props;
    return (
      <DropdownOpt
        onMouseEnter={this.setHighlighted}
        onMouseLeave={this.resetHighlighted}
        onClick={this.onClick}
        aria-selected={active}
        active={active}
        highlighted={highlighted}
        title={title}
      >
        {children}
      </DropdownOpt>
    );
  }
}
