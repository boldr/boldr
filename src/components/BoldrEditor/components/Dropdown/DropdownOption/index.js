/* @flow */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export type Props = {
  children: ?ReactChildren,
  value: ?any,
  onClick: ?Function,
  onSelect: ?Function,
  setHighlighted: ?Function,
  index: ?number,
  disabled: ?boolean,
  active: ?boolean,
  highlighted: ?boolean,
  className: ?string,
  activeClassName: ?string,
  disabledClassName: ?string,
  highlightedClassName: ?string,
  title: ?string,
};

export default class DropDownOption extends Component {
  props: Props;

  onClick: Function = (event): void => {
    const { onSelect, onClick, value, disabled } = this.props;
    if (!disabled) {
      if (onSelect) {
        onSelect(value);
      }
      if (onClick) {
        event.stopPropagation();
        onClick(value);
      }
    }
  };

  setHighlighted: Function = (): void => {
    const { setHighlighted, index } = this.props;
    setHighlighted(index);
  };

  resetHighlighted: Function = (): void => {
    const { setHighlighted } = this.props;
    setHighlighted(-1);
  };

  render(): Object {
    const {
      children,
      active,
      disabled,
      highlighted,
      className,
      activeClassName,
      disabledClassName,
      highlightedClassName,
      title,
    } = this.props;
    return (
      <li
        className={classNames('boldr-editor__dropdown-option-default', className, {
          [`boldr-editor__dropdown-option-active ${activeClassName}`]: active,
          [`boldr-editor__dropdown-option-highlighted ${highlightedClassName}`]: highlighted,
          [`boldr-editor__dropdown-option-disabled ${disabledClassName}`]: disabled,
        })}
        onMouseEnter={this.setHighlighted}
        onMouseLeave={this.resetHighlighted}
        onClick={this.onClick}
        aria-selected={active}
        title={title}
      >
        {children}
      </li>
    );
  }
}
// todo: review classname use above.
