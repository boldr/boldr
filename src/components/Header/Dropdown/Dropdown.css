.dropdown {
  z-index: 1000;
}
@media (min-width: 992px) {
  .dropdown {
    color: rgba(0,0,0,0.86);
    background-color: #fff;
    border: 1px solid rgba(0,0,0,0.1);
    border-radius: 3px;
    box-shadow: 0 2px 7px rgba(0,0,0,0.05);
    left: 20px;
    margin-top: 0;
    opacity: 0;
    position: absolute;
    top: 32px;
    transition: visibility 0s 0.2s, opacity 0.2s, margin-top 0.2s;
    visibility: hidden;
  }
}
@media (min-width: 992px) {
  .dropdown.is-open {
    margin-top: 5px;
    opacity: 1;
    transition-delay: 0s;
    visibility: visible;
  }
}

@media (min-width: 992px) {
  .highlightDropdown {
    height: 324px;
    width: 592px;
  }
}

@media (min-width: 992px) {
  .moreDropdown {
    width: 582px;
  }
}
