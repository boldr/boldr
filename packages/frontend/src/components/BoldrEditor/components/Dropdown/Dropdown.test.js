/* @flow */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Dropdown from './Dropdown';
import DropdownOption from './DropdownOption';

describe('<Dropdown />', () => {
  it('should have a div when rendered', () => {
    const wrapper = shallow(
      <Dropdown>
        <span>abcd</span>
        <DropdownOption>1234</DropdownOption>
      </Dropdown>,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
