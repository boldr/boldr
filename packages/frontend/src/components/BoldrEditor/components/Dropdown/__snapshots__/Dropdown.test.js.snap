// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should have a div when rendered 1`] = `
<Dropdown.styled__DropdownWrapper
  aria-label="be-dropdown"
  isSkinny={false}
>
  <Dropdown.styled__DropdownSelectedText>
    <span>
      abcd
    </span>
    <Dropdown.styled__CaretOpen />
  </Dropdown.styled__DropdownSelectedText>
</Dropdown.styled__DropdownWrapper>
`;
